src/vs/base/node/zip.ts

```diff
@@ -78,72 +80,88 @@ function extractEntry(stream: Readable, fileName: string, mode: number, targetPa
 		istream.once('close', () => c(null));
 		istream.once('error', e);
 		stream.once('error', e);
 		stream.pipe(istream);
-	}, () => {
-		if (istream) {
-			istream.close();
-		}
 	}));
 }
 
-function extractZip(zipfile: ZipFile, targetPath: string, options: IOptions, logService: ILogService): TPromise<void> {
+function extractZip(zipfile: ZipFile, targetPath: string, options: IOptions, logService: ILogService): CancelablePromise<void> {
 	let isCanceled = false;
-	let last = TPromise.wrap<any>(null);
+	let last = createCancelablePromise(() => Promise.resolve(null));
 	let extractedEntriesCount = 0;
 
-	return new TPromise((c, e) => {
-		const throttler = new SimpleThrottler();
+	return createCancelablePromise(token => {
+
+		once(token.onCancellationRequested)(() => {
+			logService.debug(targetPath, 'Cancelled.');
+			isCanceled = true;
+			last.cancel();
+			zipfile.close();
+		});
+
+		return new TPromise((c, e) => {
+			const throttler = new SimpleThrottler();
+
+			const readNextEntry = (token: CancellationToken) => {
+				if (token.isCancellationRequested) {
+					return;
+				}
+
+				extractedEntriesCount++;
+				zipfile.readEntry();
+			};
 
-		const readNextEntry = () => {
-			extractedEntriesCount++;
+			zipfile.once('error', e);
+			zipfile.once('close', () => last.then(() => {
+				if (isCanceled || zipfile.entryCount === extractedEntriesCount) {
+					c(null);
+				} else {
+					e(new ExtractError('Incomplete', new Error(nls.localize('incompleteExtract', "Incomplete. Found {0} of {1} entries", extractedEntriesCount, zipfile.entryCount))));
+				}
+			}, e));
 			zipfile.readEntry();
-		};
-
-		zipfile.once('error', e);
-		zipfile.once('close', () => last.then(() => {
-			if (isCanceled || zipfile.entryCount === extractedEntriesCount) {
-				c(null);
-			} else {
-				e(new ExtractError('Incomplete', new Error(nls.localize('incompleteExtract', "Incomplete. Found {0} of {1} entries", extractedEntriesCount, zipfile.entryCount))));
-			}
-		}, e));
-		zipfile.readEntry();
-		zipfile.on('entry', (entry: Entry) => {
-
-			if (isCanceled) {
-				return;
-			}
-
-			if (!options.sourcePathRegex.test(entry.fileName)) {
-				readNextEntry();
-				return;
-			}
-
-			const fileName = entry.fileName.replace(options.sourcePathRegex, '');
-
-			// directory file names end with '/'
-			if (/\/$/.test(fileName)) {
-				const targetFileName = path.join(targetPath, fileName);
-				last = mkdirp(targetFileName).then(() => readNextEntry());
-				return;
-			}
-
-			const stream = ninvoke(zipfile, zipfile.openReadStream, entry);
-			const mode = modeFromEntry(entry);
-
-			last = throttler.queue(() => stream.then(stream => extractEntry(stream, fileName, mode, targetPath, options).then(() => readNextEntry())));
+			zipfile.on('entry', (entry: Entry) => {
+
+				if (isCanceled) {
+					return;
+				}
+
+				if (!options.sourcePathRegex.test(entry.fileName)) {
+					readNextEntry(CancellationToken.None);
```

Comment:
Why are we not passing the real token instead of none token?

Reply:
Good catch, I guess I added this first and only later added the token.